SteamKit2/SteamKit2/Steam/CMClient.cs

```diff
@@ -520,40 +520,33 @@ void HandleMulti( IPacketMsg packetMsg )
             }
 
             var msgMulti = new ClientMsgProtobuf<CMsgMulti>( packetMsg );
-
-            byte[] payload = msgMulti.Body.message_body;
+            using var payloadStream = new MemoryStream( msgMulti.Body.message_body );
+            Stream stream = payloadStream;
 
             if ( msgMulti.Body.size_unzipped > 0 )
             {
-                try
-                {
-                    using var compressedStream = new MemoryStream( payload );
-                    using var gzipStream = new GZipStream( compressedStream, CompressionMode.Decompress );
-                    using var decompressedStream = new MemoryStream( capacity: ( int )msgMulti.Body.size_unzipped );
-                    gzipStream.CopyTo( decompressedStream );
-                    payload = decompressedStream.ToArray();
-                }
-                catch ( Exception ex )
-                {
-                    LogDebug( "CMClient", "HandleMulti encountered an exception when decompressing.\n{0}", ex.ToString() );
-                    return;
-                }
+                stream = new GZipStream( payloadStream, CompressionMode.Decompress );
             }
 
-            using var ms = new MemoryStream( payload );
-            using var br = new BinaryReader( ms );
-            while ( ( ms.Length - ms.Position ) != 0 )
+            using ( stream )
             {
-                int subSize = br.ReadInt32();
-                byte[] subData = br.ReadBytes( subSize );
+                Span<byte> length = stackalloc byte[ sizeof( int ) ];
 
-                if ( !OnClientMsgReceived( GetPacketMsg( subData, this ) ) )
+                while ( stream.Read( length ) == length.Length )
```

Comment:
This is slightly sketchy, I don't know if there's a better way to know if gzip stream has reached the end.

Reply:
Isn't the standard contract `while (bytes_read > 0)`?

I think here it would be something like:

```cs
while ( stream.Read( length ) is > 0 and var dataLength )
{
    var data = length.Slice(0, dataLength);
    // ..
}
```

The problem there is you might not get the full set of length bytes in one read.

You might need to loop over it, subtracting any bytes read in a previous partial read?